peak/mapper/binding.py

```diff
@@ -0,0 +1,195 @@
+import typing as tp
+import itertools as it
+from hwtypes import SMTBit, SMTBitVector, SMTSIntVector
+from hwtypes import is_adt_type
+from hwtypes.adt import Product, Enum
+
+def is_product(isa):
+    return issubclass(isa,Product)
+
+#finds all paths in the adt
+#A path is a tuple of names that indicate location in nested Product
+#TODO does not deal with Sum (treats it as a leaf type)
+def _flatten_adt(isa,path=()) -> tp.Mapping[tuple,type]:
+    if issubclass(isa,Product):
+        res = {}
+        for name,t in isa.field_dict.items():
+            res.update(_flatten_adt(t,path+(name,)))
+        return res
+    else:
+        return {path:isa}
+
+def _sort_by_t(path2t : tp.Mapping[tuple,type]) ->tp.Mapping[type,tp.List[tuple]]:
+
+    t2path = {}
+    for tup,t in path2t.items():
+        t2path.setdefault(t, []).append(tup)
+
+    return t2path
+
+#constructs a default instruction
+def _default_instr(isa,forall=False):
+    if issubclass(isa,Product):
+        return isa(**{name:_default_instr(t,forall) for name,t in isa.field_dict.items()})
+    elif issubclass(isa,Enum):
+        return isa.fields[0]
+    elif forall:
+        return isa()
+    else:
+        return isa(0)
+
+def _get_from_path(instr,path):
+    if path is ():
+        return instr
+    else:
+        assert isinstance(instr,Product)
+        return _get_from_path(getattr(instr,path[0]),path[1:])
+
+def _set_from_path(instr,path,val):
+    setattr(_get_from_path(instr,path[:-1]),path[-1],val)
+
+
+#Small class that only creates new free smt vars when it needs to
+class SMT_cache:
+    def __init__(self):
+        self.cache = {}
+        self.idx = {}
+
+    def reset(self):
+        self.idx = {}
+
+    def __getitem__(self,t):
+        idx = self.idx.setdefault(t,0)
+        if not t in self.cache:
+            self.cache[t] = [t()]
+        if len(self.cache[t])==idx:
+            self.cache[t].append(t())
+        val = self.cache[t][idx]
+        self.idx[t] +=1
+        return val
```

Comment:
This seems to me like it just going to create a new variable every time...

On first run get:
t not in idx, idx := 0
t not in cache, cache[t] := [t()]
len cache != idx, ...
val := cache[t][0]
idx[t] := 1

on second run:
idx := 1
len cache == 1, cache.append(t())
val := cache[t][1]
idx[t] := 2

this only stops on reset when old variables will be returned but constructing a new variable isn't expensive so I don't see what we are gaining from this.

Reply:
Removing